File: cli/src/commands/session/handoff.ts
import { Command } from "commander";
import { daemonPost } from "../../lib/daemon-client.js";
import { preflight } from "../../lib/preflight.js";
import { die, success } from "../../lib/output.js";

interface HandoffResponse {
  ok: true;
  handoffSummary: string | null;
}

export function registerSessionHandoff(session: Command): void {
  session
    .command("handoff <id>")
    .description("Generate a handoff summary for a session")
    .action(async (id: string) => {
      await preflight();

      const result = await daemonPost<HandoffResponse>(
        `/sessions/${encodeURIComponent(id)}/handoff`
      );

      if (!result.ok) {
        die(result.error, result.status === 404 ? 2 : 1);
      }

      if (result.data.handoffSummary) {
        success(result.data.handoffSummary);
      } else {
        success("No handoff summary was produced.");
      }
    });
}

File: rust/vst-cli/src/commands/agent/handoff.rs
use vst_types::rest::sessions::HandoffResult;

use crate::client::{daemon_post, DaemonResult};
use crate::output::success;
use crate::preflight::preflight;

pub async fn run_session_handoff(id: &str) -> Result<(), (String, i32)> {
    if id.is_empty() {
        return Err(("Session ID is required".to_string(), 1));
    }

    preflight().await;

    let mut encoded = String::new();
    for byte in id.bytes() {
        match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                encoded.push(byte as char);
            }
            _ => {
                encoded.push_str(&format!("%{byte:02X}"));
            }
        }
    }

    let result = daemon_post::<HandoffResult, ()>(&format!("/sessions/{encoded}/handoff"), None)
        .await
        .map_err(|e| (e.to_string(), 1))?;

    match result {
        DaemonResult::Ok { data, .. } => {
            if let Some(summary) = data.handoff_summary {
                success(&summary);
            } else {
                success("No handoff summary was produced.");
            }
            Ok(())
        }
        DaemonResult::Err { status, error, .. } => {
            let code = if status == 404 { 2 } else { 1 };
            Err((error, code))
        }
    }
}
